src/main/java/com/fasterxml/jackson/databind/deser/BeanDeserializerBase.java

```diff
@@ -212,7 +212,25 @@ protected BeanDeserializerBase(BeanDeserializerBuilder builder,
         _injectables = (injectables == null || injectables.isEmpty()) ? null
                 : injectables.toArray(new ValueInjector[injectables.size()]);
         _objectIdReader = builder.getObjectIdReader();
+
+        // If you check for canCreateFromInt() below which appears to be the equivalent
+        // of the existing canCreateUsingArrayDelegate() it also causes the included
+        // test testPOJOWithPrimitiveCreatorFromObjectRepresentation() to break in the
+        // same way as the test testPOJOWithArrayCreatorFromObjectRepresentation() which
+        // I assert is failing because of the call to canCreateUsingArrayDelegate() below.
+        //
+        // Removing the call to canCreateUsingArrayDelegate() below allows the test
+        // testPOJOWithArrayCreatorFromObjectRepresentation() to pass and interestingly
+        // all existing tests continue to pass. I somehow doubt that it's that simple.
+        //
+        // Specifically, it seems to me that the vanilla-ness of the deserialization can
+        // only be determined once the structure of the incoming JSON can be compared to
+        // what creator methods exist. For example, if there is an array delegate and we
+        // are asked to deserialize from an array then it's not vanilla. However, if we
+        // are asked to deserialize from an object then it's vanilla (subject to no other
+        // non-vanilla-ness being present).
```

Comment:
Quick note here: vanilla-ness and non-standard creation are not well documented, and definition is purely practical to try to allow simpler looping methods to be called if (and only if) they can be guaranteed to be sufficient.
I _think_ (but will need to double-check) that omission of scalar creators was intentional, but should be mentioned in a comment.

Vanilla processing, for example, only affects things in case where POJO-style mapping of properties from JSON Object is done... it is all quite confusing. I guess another way to put this is that yes, "total vanilla-ness" is combination of incoming shape and other checks that can be calculated a priori, from static configuration.
So flag is that partial answer: "yes, we already know there are special cases to check".

Especially since addition of "array delegate" was an external contribution after the fact and deviates from original ideas a bit.

Anyway: I will merge this, have a look.